dbt/adapters/snowflake/connections.py

```diff
@@ -435,32 +436,43 @@ def add_query(self, sql, auto_begin=True, bindings=None, abridge_sql_log=False):
             bindings = tuple(bindings)
 
         queries = self._split_queries(sql)
-
-        for individual_query in queries:
+        stripped_queries = []
+        for query in queries:
             # hack -- after the last ';', remove comments and don't run
             # empty queries. this avoids using exceptions as flow control,
             # and also allows us to return the status of the last cursor
             without_comments = re.sub(
                 re.compile(r"(\".*?\"|\'.*?\')|(/\*.*?\*/|--[^\r\n]*$)", re.MULTILINE),
                 "",
-                individual_query,
+                query,
             ).strip()
 
-            if without_comments == "":
-                continue
+            if without_comments != "":
+                stripped_queries.append(query)
 
+        for individual_query in stripped_queries:
             # Even though we turn off transactions by default for Snowflake,
             # the user/macro has passed them *explicitly*, probably to wrap a DML statement
-            # Let their wish be granted!
+            # Assuming this BEGIN/COMMIT is not all alone - let their wish be granted!
             # This also has the effect of ignoring "commit" in the RunResult for this model
             # https://github.com/dbt-labs/dbt-snowflake/issues/147
-            if individual_query.lower() == "begin;":
-                super().add_begin_query()
-                continue
+            # (Otherwise, if this 'BEGIN' or 'COMMIT' query *is* all by itself, let it be run as a standalone query)
+            if len(stripped_queries) == 1 and individual_query.lower() in ("begin;", "commit;"):
```

Comment:
I think this check should happen outside of the for loop. Basically "if you handed me a single commit; or begin;, then warn. otherwise start processing some queries".

Reply:
@mikealfare There's some trickiness here. If we're handed a single query AND that query is `begin;` or `commit;`, we want to pass it through, to avoid this error:
```
[DEBUG] stdout_log: 16:55:24.961406 [debug] [Thread-1  ]: Runtime Error in model my_model (models/my_model.sql)
  Tried to run an empty query on model 'model.test.my_model'. If you are conditionally running
  sql, eg. in a model hook, make sure your `else` clause contains valid sql!

  Provided SQL:

      begin;
```
But if we're handed _multiple_ queries (semicolon-delimited), which are wrapped in a `begin;` and a `commit;`, then we want to do the handling below, that converts them into `super().add_begin_query()` and `super().add_commit_query()`. (This was the fix for https://github.com/dbt-labs/dbt-snowflake/issues/147).

So it is possible to pull this warning out of the `for` loop, but we still need an added conditional check within the `for` loop. Do you prefer it being written this way?
```python
        # Even though we turn off transactions by default for Snowflake,
        # the user/macro has passed them *explicitly*, probably to wrap a DML statement.
        # Assuming this BEGIN/COMMIT is not all alone - let their wish be granted!
        # This also has the effect of ignoring "commit" in the RunResult for this model
        # https://github.com/dbt-labs/dbt-snowflake/issues/147
        
        # (However, if this 'BEGIN' or 'COMMIT' query *is* all by itself, raise a warning and let it be run as a standalone query)
        if len(stripped_queries) == 1 and stripped_queries[0].lower() in ("begin;", "commit;"):
            message = (
                "Explicit transactional logic should be used only to wrap "
                "DML logic (MERGE, DELETE, UPDATE, etc). The keywords BEGIN; and COMMIT; should "
                "be placed directly before and after your DML statement, rather than in separate "
                "statement calls or run_query() macros."
            )
            logger.warning(line_wrap_message(warning_tag(message)))

        for individual_query in stripped_queries:
            if len(stripped_queries) > 1 and individual_query.lower() == "begin;":
                super().add_begin_query()
                continue

            elif len(stripped_queries) > 1 and individual_query.lower() == "commit;":
                super().add_commit_query()
                continue
```

Reply:
I see what you're saying. So then you'd need to put the rest in an `else` and add a step to the "exception" condition. That would also get rid of the need for `continue`, which improves readability in my opinion. Basically I'd rather separate the "exception" state (a single `begin;`/`commit;`) from the "happy path"/"normal" state. So it could look like this:
```
for query in queries:
    ...

def add_standard_query(query_def):
    return super().add_query(
        self._add_query_comment(query_def),
        auto_begin,
        bindings=bindings,
        abridge_sql_log=abridge_sql_log
    )

if len(stripped_queries) == 1 and stripped_queries[0].lower() in ["begin;", "commit;"]:
    # comments speaking to exception case
    message = (...)
    logger.warning(line_wrap_message(warning_tag(message)))
    connection, cursor = add_standard_query(stripped_queries[0])
else:
    for query in queries:
        # comments, if any, speaking to happy path
        if query.lower() == "begin;":
            super().add_begin_query()
        elif query.lower() == "commit;":
            super().add_commit_query()
        else:
            connection, cursor = add_standard_query(query)

if cursor is None:
    ...

return connection, cursor
```

It appears that the last exception logic is only triggered if we receive a series of `begin;` or `commit;` queries. So our three scenarios are:

1. Single `begin;` or `commit;` > let it pass through as a standard query with no error
2. A series of queries which are all either `begin;` or `commit;` > throw a DbtRuntimeError
3. A series of queries, at least one of which is not `begin;` or `commit;` > the cursor is set

If that's the case, should we combine the first two scenarios? It could look like this:
```
for query in queries:
    ...

def add_standard_query(query_def):
    return super().add_query(
        self._add_query_comment(query_def),
        auto_begin,
        bindings=bindings,
        abridge_sql_log=abridge_sql_log
    )

if set(query.lower() for query in stripped_queries).issubset(set("begin;", "commit;")):
    # if all we get are a series of `begin;` and/or `commit;`, return one of them as a standard query to avoid an error downstream
    # comments speaking to exception case
    message = (...)
    logger.warning(line_wrap_message(warning_tag(message)))
    return add_standard_query(stripped_queries[0])  # do we even want to return a query?

for query in queries:
    # comments, if any, speaking to happy path
    if query.lower() == "begin;":
        super().add_begin_query()
    elif query.lower() == "commit;":
        super().add_commit_query()
    else:
        connection, cursor = add_standard_query(query)

return connection, cursor
```

Reply:
@mikealfare I like it! Much cleaner separation of the logic. I just pushed up a change in this vein, let me know what you think

Reply:
Two items, seen below. I think there's a missing `return`, and we implemented something similar to the `add_standard_query()` method in `dbt-core`.